### DISCUSSION TITLE
Rethinking the handling of commands

### DISCUSSION THREAD

**Discussion by lefou** [com-lihaoyi/mill#2594] [Ideas]
This is the attempt to persist an interesting conversation on Discord and continue it.

Here is the conversation:

@lefou wrote:

> Wouldn't it be nice, to get the task-part of a command without the command-part, and use it like a  task?
>My typical pattern is currently, to write a task which I then use in a command.
>
> ```scala
> def fooTask(x: X): Task[Y] = T.task {
>   ???
> }
> def foo(x: X): Command[Y] = T.command {
>   fooTask(x)
> }
> ```

@lihaoyi wrote:

> i was thinking that we could possibly just treat commands as tasks when used in other targets, e.g. inheriting the working directory rather than using their own (since it has no longer any guarantee of being unique)

I'd like to add, that there are other issue with command, e.g. we don't have a nice compatibility story to evolve existing commands. As `Command`s need to be resolved via reflection and also support CLI options parsing, we can't rely on Scala `def` overloading. Hence, it's not easy to evolve a command that accepts `args: String*` to a more type-safe API.

But we are more free when it comes to `Task`s. We could easily add a new tasks accepting more arguments and just forward the command to that task with some mainargs magic applied.

What, if we write all commands as tasks and then only add a small marker to the tasks that should be exposed as a command? 

I envision some more binary compatible signature for commands, e.g. `Array[String]` like we have for `def main`. 

There are multiple ways to realize that. `T.command` could be just an alias to `T.task` plus the generation of a synthetic command. Or some other way. I have not really thought it through, especially regarding what's possible with macros in Scala 2.13 vs Scala 3. Hence, **I'd like to hear your comments and ideas!** Hopefully, we can come up with an design overhault and a final pull request.

**Comment by lefou** [comment]
Related: 
* https://github.com/com-lihaoyi/mill/issues/4854#issuecomment-2776096063 and following

https://github.com/com-lihaoyi/mill/issues/4854#issuecomment-2776096063
> So, it should work in our normal sources as well. This looks like a good pattern to use in Mill. Maybe we can even drop some boilerplate. 
>
> ```scala
>case class MyCmdConfig(name: String)
>
>def cm1(config: MyCmdCOnfig) = Command[MyCmdConfig] {
>  println("hello " + config.name
>}
>
>// or
>
>def cmd2(args: String*) = Task.Command[MyCmdConfig] { config =>
>  println("hello " + config.name)
>}
>```
>
> Question is, what is best suited to later change without breakage. From compatibility the String-version is better, as it should even work with a complete different config object when overridden. OTOH, when we want to call it from code, the typed version is better. But we seldomly want to call commands from code.

https://github.com/com-lihaoyi/mill/issues/4854#issuecomment-2776104953
>Ideally, we can get a typed anonymous task suited to be called from code and a untyped command just mapping that task.
>
>```scala
>def cmdTask(config: MyCmdCOnfig) = Task.Anon[MyCmdConfig] {
>  println("hello " + config.name)
>}
>
>def cmd(args: String*) = Task.Command(cmdtask)
>```
>
>or 
>
>
>```scala
>def cmdTask(config: MyCmdCOnfig) = Task.Anon {
>  println("hello " + config.name)
>}
>
>def cmd(args: String*) = Task.Command[MyCmdConfig](config => cmdtask(config))
>// or
>def cmd(args: String*) = Task.Command[MyCmdConfig](cmdtask)
>```

**Reply by lefou** [reply]
Although it would be superior in terms of compatibility, having three definitions (the case class in the companion object, the anon task and the command) is much more boilerplate.

**Comment by lihaoyi** [comment]
In theory the `case class` approach and the flat-parameter-list approach should have the same compatibility characteristics. In both cases, evolution can be handled by the `@unroll` annotation or by manual definition of forwarders

In practice that might not be true right now due to edge cases in the macros and type inference, but it should be possible to make it work

**Comment by lefou** [comment]
Every now an then I hit this command issues again. And it always means, I can't change it (due to compat constraints), although there are probably no API clients, only CLI clients. I think the `String`-based definition would be the best approach, since it can be applied within the current binary comaptibility window. It's also the form from which we often want to evolve from. Also, completely changing the comand signature would be possible in the same way we can change `@main` entry point to an app, since we can softly "remove" parameters.

**Reply by lefou** [reply]
We want:
* a `String*`-based signature - so it can evolve freely
* out-of-the-box type-safe argument parsing
* support for usage rendering via `inspect`

```scala
case class MyCmdConfig(...)

def cmd(args: String*) = Task.Command {
  // use the raw args, e.g. parse manually
  val config = mainargs.Parser[MyCmdConfig].constructOrExit(args)
}

def cmd(args: String*) = Task.Command { (config: MyCmdConfig) =>
  // use type-safe args
}
```
